From 76993440d38f2d12fb7d8f0f6ec069f93eb6a648 Mon Sep 17 00:00:00 2001
From: Jai Luthra <j-luthra@ti.com>
Date: Mon, 28 Aug 2023 18:18:39 +0530
Subject: [PATCH 002/322] media: ti: j721e-csi2rx: Assert pixel reset before
 stopping last stream

To prevent stale data in CSI2RX FIFOs ensure that pixel reset is
asserted before stopping last stream. Not doing so prevents module
unload/reload.

Signed-off-by: Jai Luthra <j-luthra@ti.com>
Signed-off-by: Vignesh Raghavendra <vigneshr@ti.com>
---
 .../media/platform/ti/j721e-csi2rx/j721e-csi2rx.c  | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/drivers/media/platform/ti/j721e-csi2rx/j721e-csi2rx.c b/drivers/media/platform/ti/j721e-csi2rx/j721e-csi2rx.c
index bfd35d75f80d..c1213d381718 100644
--- a/drivers/media/platform/ti/j721e-csi2rx/j721e-csi2rx.c
+++ b/drivers/media/platform/ti/j721e-csi2rx/j721e-csi2rx.c
@@ -589,6 +589,12 @@ static void ti_csi2rx_setup_shim(struct ti_csi2rx_ctx *ctx)
 		return;
 	}
 
+	/* De-assert the pixel interface reset. */
+	if (!csi->enable_count) {
+		reg = SHIM_CNTL_PIX_RST;
+		writel(reg, csi->shim + SHIM_CNTL);
+	}
+
 	reg = SHIM_DMACNTX_EN;
 	reg |= FIELD_PREP(SHIM_DMACNTX_FMT, fmt->csi_dt);
 
@@ -1022,6 +1028,10 @@ static void ti_csi2rx_stop_streaming(struct vb2_queue *vq)
 	struct ti_csi2rx_dev *csi = ctx->csi;
 	int ret;
 
+	/* assert pixel reset to prevent stale data on stopping last stream */
+	if (csi->enable_count == 1)
+		writel(0, csi->shim + SHIM_CNTL);
+
 	video_device_pipeline_stop(&ctx->vdev);
 
 	ret = v4l2_subdev_disable_streams(&csi->subdev,
@@ -1638,10 +1648,6 @@ static int ti_csi2rx_probe(struct platform_device *pdev)
 		goto cleanup_subdev;
 	}
 
-	/* De-assert the pixel interface reset. */
-	reg = SHIM_CNTL_PIX_RST;
-	writel(reg, csi->shim + SHIM_CNTL);
-
 	return 0;
 
 cleanup_subdev:
-- 
2.43.0

